projects/packages/waf/src/class-waf-runtime.php

```diff
@@ -279,6 +318,20 @@ public function match_targets( $transforms, $targets, $match_operator, $match_va
 	 * @param int    $status_code Http status code.
 	 */
 	public function block( $action, $rule_id, $reason, $status_code = 403 ) {
+		if ( 'ip block list' === $reason ) {
+			$real_ip = $this->request->get_real_user_ip_address();
+
+			if ( $this->is_ip_allowed_for_recovery( $real_ip ) ) {
+				return;
+			}
+
+			global $pagenow;
+			if ( isset( $pagenow ) && 'wp-login.php' === $pagenow ) {
+				$this->allow_login_or_prompt_recovery( $real_ip );
+				return;
+			}
+		}
```

Comment:
Wonder if we could have this logic in a separate function, and implement it higher up so we can have a nice looking check like:

```
if ( should_block_with_recovery() ) {
    $blocked_login_page->block();
} else {
   $waf->block();
}
```

Reply:
I believe the actual use of `$waf->block` is written right into the rules file in `Waf_Rules_Manager`, I'd agree that this would be an ideal set up but I am not sure there is a "higher up" placement unfortunately.